src/store.rs

```diff
@@ -1,54 +1,77 @@
 use bytes::Bytes;
-use std::collections::HashMap;
+use std::collections::{BTreeSet, HashMap};
 use std::ops::AddAssign;
+use std::ops::Deref;
 use std::str::FromStr;
+use std::sync::{Arc, Mutex, MutexGuard};
+use tokio::sync::Notify;
+use tokio::time::{sleep_until, Duration, Instant};
 
+#[derive(Clone)]
 pub struct Store {
-    store: HashMap<String, Bytes>,
+    inner: Arc<InnerStore>,
 }
 
 impl Store {
     pub fn new() -> Store {
-        Store {
-            store: HashMap::new(),
-        }
-    }
+        let state = State {
+            keys: HashMap::new(),
+            ttls: BTreeSet::new(),
+        };
 
-    pub fn set(&mut self, key: String, value: Bytes) {
-        self.store.insert(key, value);
-    }
+        let waker = Notify::new();
+        let inner = Arc::new(InnerStore {
+            state: Mutex::new(state),
+            waker,
+        });
 
-    pub fn get(&self, key: &str) -> Option<&Bytes> {
-        self.store.get(key)
-    }
+        tokio::spawn({
+            let inner = inner.clone();
+            async move { remove_expired_keys(inner).await }
+        });
 
-    pub fn remove(&mut self, key: &str) -> Option<Bytes> {
-        self.store.remove(key)
+        Self { inner }
     }
+}
 
-    pub fn exists(&self, key: &str) -> bool {
-        self.store.contains_key(key)
+impl Default for Store {
+    fn default() -> Self {
+        Self::new()
     }
+}
 
-    pub fn keys(&self) -> impl Iterator<Item = &String> {
-        self.store.keys()
+pub struct InnerStore {
+    state: Mutex<State>,
+    waker: Notify,
+}
+
+impl Deref for Store {
+    type Target = InnerStore;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
     }
+}
 
-    pub fn size(&self) -> usize {
-        self.store.len()
+impl InnerStore {
+    pub fn lock(&self) -> MutexGuard<State> {
+        self.state.lock().unwrap()
     }
 
-    pub fn iter(&self) -> impl Iterator<Item = (&String, &Bytes)> {
-        self.store.iter()
+    pub fn set2(&self, key: Key, value: NewValue) {
+        let mut state = self.lock();
+        state.set2(key, value);
+        self.waker.notify_one();
```

Comment:
[Q] What is notify accomplishing here?

Reply:
If we insert a new key where the expires_at attribute would be the next to expire, we need to notify the background job to recognize this key as the next expiring key.

Reply:
We should then only notify if there's an expiration set, to avoid unnecessary calls.

Reply:
Yes, indeed. We could even optimize by checking if the newly inserted ttl is the next expiring key, and if not, skip the notification.